### PR TITLE
Issue #509 Add developer User to sudoers

### LINKED ISSUE #509: Give the developer sudoers role
Recently @hferentschik pointed a nice article about how we can implement same linux philosophy of having sudo to perform admin tasks and I think this will helpful for us and for users.

- https://bierkowski.com/openshift-cli-morsels-sudo-and-user-impersonation/

### PR DESCRIPTION
**Pull request by praveenkumar**
Fix #509

### CODE REVIEW

File: pkg/minishift/openshift/oc.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+package openshift
+
+import (
+	instanceState "github.com/minishift/minishift/pkg/minishift/config"
+	"github.com/minishift/minishift/pkg/util"
+)
+
+// runner executes commands on the host
+var runner util.Runner = &util.RealRunner{}
```
Comment (budhrg) [inline pkg/minishift/openshift/oc.go]: `util.Runner` is it required? 

I think datatype will be inferred here.

File: test/integration/util/util.go [hunk 2]
```diff
@@ -120,7 +122,7 @@ type OcRunner struct {
 }
 
 func NewOcRunner() *OcRunner {
-	p, _ := exec.LookPath("oc")
```
Comment (praveenkumar) [inline test/integration/util/util.go]: This will not work because as per doc LookPath only able to search binary if that is part of $PATH https://golang.org/pkg/os/exec/#example_LookPath

File: test/integration/util/util.go [hunk 3]
```diff
@@ -120,7 +122,7 @@ type OcRunner struct {
...
+	p := path.Join(os.Getenv(constants.MiniShiftHomeEnv), "cache", "oc", "v1.4.1", "oc")
```
Comment (budhrg) [inline test/integration/util/util.go]: :+1:

File: pkg/minishift/openshift/oc.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	cmdArgs = []string{"adm", "policy", "add-cluster-role-to-user", "sudoer", "developer"}
+	if _, err := runner.Output(cmdName, cmdArgs...); err != nil {
+		return err
+	}
+	cmdArgs = []string{"login", "-u", "developer"}
+	if _, err := runner.Output(cmdName, cmdArgs...); err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (budhrg) [inline pkg/minishift/openshift/oc.go]: go fmt will fail here.

File: pkg/minishift/openshift/oc.go (hunk 1 above)
Comment (praveenkumar) [inline pkg/minishift/openshift/oc.go]: @budhrg without runner we can't execute the command to system (VM)

Comment (budhrg) [inline pkg/minishift/openshift/oc.go]: Sorry, should have commented clearly.

I mean to say this `var runner = &util.RealRunner{}` , We can omit datatype during variable definition.

File: test/integration/util/util.go (hunk 3 above)
Comment (praveenkumar) [inline test/integration/util/util.go]: This is still hard code the oc version so I think will use `version.GetOpenShiftVersion()`

File: pkg/minishift/openshift/oc.go (hunk 1 above)
Comment (praveenkumar) [inline pkg/minishift/openshift/oc.go]: @budhrg right but this will make little bit clear where actually it's coming from but yes we can omit.

File: test/integration/util/util.go (hunk 3 above)
Comment (praveenkumar) [inline test/integration/util/util.go]: Looks like `version.GetOpenShiftVersion()` not going to work because it will going to get `v0.0.0-unset` as version.

Comment (budhrg) [inline test/integration/util/util.go]: Can we at least extract oc version into some constants ? 

We eventually might need to extract common test constants to some file

Comment (praveenkumar) [inline test/integration/util/util.go]: @budhrg I think best way to handle it during setup add the `testDir` to $PATH and then use `LookPath` but let's discuss it.

Comment (budhrg) [inline test/integration/util/util.go]: @praveenkumar , 
I think we can cache the common dirs like `oc` dir, as 

```
OcDir := path.Join(os.Getenv(constants.MiniShiftHomeEnv), "cache", "oc")
ISODir := path.Join(os.Getenv(constants.MiniShiftHomeEnv), "cache", "iso")
```

etc.

Doesn't feel good to add testDir to $PATH. This also might not work in Windows but unsure.

Comment (budhrg) [inline test/integration/util/util.go]: Not exactly what I thought. What if I had to use this path in any other code? 

I think let's not do this as part of it. We can refactor it later or discuss the proper solution.

File: test/integration/features/basic.feature [hunk 5]
```diff
@@ -7,6 +7,7 @@ Feature: Basic
     Given executing "minishift start --docker-env=FOO=BAR --docker-env=BAZ=BAT"
      Then Minishift should have state "Running"
       And Minishift should have a valid IP address
+      And Developer should have sudoers access
```
Comment (gbraad) [inline test/integration/features/basic.feature]:
```gherkin
When executing "openshift fdfsfd"  # or similar
Then developer should have sudo access
```

You intertwine the cases of action and behaviour

File: test/integration/basic_test.go [hunk 6]
```diff
@@ -66,6 +66,16 @@ func (m *Minishift) shouldHaveAValidIPAddress() error {
 	return nil
 }
 
+func shouldHaveSudoerAccess() error {
+	ocRunner := util.NewOcRunner()
+	cmdArgs := []string{"--as", "system:admin", "get", "node"}
+	cmdOut, _ := ocRunner.RunCommand(cmdArgs)
+	if !strings.Contains(string(cmdOut), "NAME") {
+		return fmt.Errorf("Not able to set developers as sudoers: %s", string(cmdOut))
```
Comment (gbraad) [inline test/integration/basic_test.go]: maybe I am not following your intention here. likely due to the fact of the error message.

Does it set the access level, or perform an action as a certain user?

**Comment by budhrg** [comment]
@praveenkumar , Could we have commit message in this format `Issue #<issue no> message`? I mean `#` is missing.

File: test/integration/features/basic.feature [hunk 7]
```diff
@@ -28,4 +36,4 @@ Feature: Basic
      When executing "minishift stop"
      Then Minishift should have state "Stopped"
      When executing "minishift delete"
-     Then Minishift should have state "Does Not Exist"
+     Then Minishift should have state "Does Not Exist"
```
Comment (gbraad) [inline test/integration/features/basic.feature]: newline or not? I get confused what people want...

File: test/integration/features/basic.feature [hunk 8]
```diff
@@ -8,6 +8,14 @@ Feature: Basic
      Then Minishift should have state "Running"
       And Minishift should have a valid IP address
```
Comment (gbraad) [inline test/integration/features/basic.feature]: It is probably better to actually capture this in a new function that describes more in detail what happens... and why this output. but for the sake of the PR, it might be an idea to just add a comment above:

```gherkin
  # Check if 
  When executing "oc ..."
```

File: test/integration/features/basic.feature (hunk 7 above)
Comment (praveenkumar) [inline test/integration/features/basic.feature]: @gbraad It's not the new line it just that I put my features bit before it so that's comes it like that. I don't like this way so showing stuff from github (it's confusing) :(

File: test/integration/features/basic.feature (hunk 8 above)
Comment (praveenkumar) [inline test/integration/features/basic.feature]: @gbraad I agree, I added some comment so that we are clear what we want to achieve here.

File: test/integration/features/basic.feature (hunk 7 above)
Comment (praveenkumar) [inline test/integration/features/basic.feature]: @gbraad nvm I got what you wanted to say. I added newline to EOF.

Comment (gbraad) [inline test/integration/features/basic.feature]: Check...

but we can get the IP of the VM. this is `minishift ip`, and we can actually store this if needed... we will probably need this more often. But let's leave this as a possible improvement.

We could collect a list for this in a new issue.

File: test/integration/features/basic.feature [hunk 9]
```diff
@@ -8,6 +8,19 @@ Feature: Basic
...
 
+     # Check if developer have sudo permission then oc should output as below
+     # Since we can't determine the vm ip before had so just make sure `Name`
+     # should be part of output string
+     # NAME           STATUS    AGE
+     # 192.168.64.2   Ready     1m
+     When executing "oc --as system:admin get node"
```
Comment (hferentschik) [inline test/integration/features/basic.feature]: This is really a indirect test. It makes several assumptions which are not clear, for example that a non sudoer user cannot view this. I think we should add a couple of things here. The direct test imo is `oc get cluster role sudoer` needs to return sudoer. 

If we want we can than also keep this test, but would also add the opposite. `When executing "oc  get node" where the output is "User "developer" cannot list all nodes in the cluster" and the exit status 1

File: test/integration/util/util.go (hunk 3 above)
Comment (hferentschik) [inline test/integration/util/util.go]: I don't get it. Why are you not using '~/.minishift/machines/minishift.json'? That's why we introduced it, right? The LookupPath was a crutch until we could do it. 

So as a minimum the oc path should be read from the config file. 

There are a couple of other issues here which need addressing. Preferably sooner than later. 

* The integration tests are interfering atm with the default Minishift directory. As part of the integration test setup we should create a known directory within our source tree (e.g. out/integration-test or something). Easiest would be to create the directory in make and then set this directory as value for MINISHIFT_HOME when calling the integration test. Problem is that in this case one needs to drive the test via Make. 
* The second problem is constants.MachineName, we need to take this value from an env variable as well, otherwise even though we can separate the home directories, we still cannot create multiple machines since the names are clashing. It needs to be tested whether the name can easily change. There might be some hard coded values which will cause trouble.

Bottom line, as a bare minimum we need to adjust the path. I'd love to see the other two issues addressed as well, but let's discuss.

File: test/integration/util/util.go [hunk 10]
```diff
@@ -34,9 +36,9 @@ type MinishiftRunner struct {
 	CommandArgs string
 }
 
-func (m *MinishiftRunner) RunCommand(command string) (stdOut string, stdErr string, exitCode int) {
+func runCommand(command string, commandPath string) (stdOut string, stdErr string, exitCode int) {
```
Comment (hferentschik) [inline test/integration/util/util.go]: hmm, I'd probably wrap this into a struct similar to _MinishiftRunner_, just _OcRunner_ and then I let me give _MinishiftRunner_ an instance of _OcRunner_.

File: test/integration/util/util.go [hunk 11]
```diff
@@ -120,7 +127,7 @@ type OcRunner struct {
 }
 
 func NewOcRunner() *OcRunner {
```
Comment (hferentschik) [inline test/integration/util/util.go]: Maybe this constructor should actually take the path. This way it becomes more flexible. One can determine the right path from the outside and create a Runner with any odd oc version, not just the one provided by minishift.json

File: pkg/minishift/openshift/oc.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import (
+	instanceState "github.com/minishift/minishift/pkg/minishift/config"
+	"github.com/minishift/minishift/pkg/util"
+)
+
+// runner executes commands on the host
+var runner util.Runner = &util.RealRunner{}
+
+// Add developer user to cluster sudoers
+func AddSudoersRoleForDeveloperUser() error {
```
Comment (hferentschik) [inline pkg/minishift/openshift/oc.go]: Looking into the future something like _AddSudoersRoleForUser(user string)_ might be better. Once we add the ability to create users, we probably want to add sudoer roles fro them as well.

File: test/integration/util/util.go (hunk 2 above)
Comment (hferentschik) [inline test/integration/util/util.go]: This was a crutch until we would be able to tell which oc version to use. We can now.

**Comment by hferentschik** [comment]
Ahh, and this really need to be mentioned in the docs somewhere. Probably somewhere were we talk about the various logins.

**Comment by budhrg** [comment]
Hi @Preeticp , I think we don't add labels to PR so I have removed it. Added doc label to related issue. Not sure about needs-investigation label.

**Comment by Preeticp** [comment]
@budhrg thanks for moving the tags to the issue.
Afaiu, 'needs-investigation' tag is for tasks which are still being investigated and worked on by the developers, and I figured work is still happening on this PR. If dev work on this PR is complete, then we do not need the 'needs-investigation' tag.

**Comment by budhrg** [comment]
> Afaiu, 'needs-investigation' tag is for tasks which are still being investigated and worked on by the developers

@Preeticp , actually `needs-investigation` is to be added to issues which developers think that they can't able to complete at the moment probably due to incomplete knowledge and think if done bit of investigation then can able to complete the task. Example, https://github.com/minishift/minishift/issues/529 where we found slow communication when using Minishift + Hyper-V. Now, at least for me I don't know much the internals of hyper-v hypervisor, and need to investigate why this behavior particularly for Hyper-V.

Good to know you are also taking part in labelling :+1:

**Comment by hferentschik** [comment]
Actually, I think we are talking about just a couple of sentences here and maybe a link to some relevant information. @praveenkumar, can you not add this so that we can merge this asap.

**Comment by praveenkumar** [comment]
> can you not add this so that we can merge this asap.

@hferentschik I hit with some of different issue during the change, will able to push changes as per your suggestion soon

File: test/integration/basic_test.go [hunk 13]
```diff
@@ -154,7 +171,7 @@ func FeatureContext(s *godog.Suite) {
 
 	s.BeforeScenario(func(interface{}) {
 		testDir := setUp()
-		os.RemoveAll(testDir)
```
Comment (praveenkumar) [inline test/integration/basic_test.go]: This is no use in our case and shouldn't be part of BeforeScenario.

File: test/integration/basic_test.go [hunk 14]
```diff
@@ -154,7 +171,7 @@ func FeatureContext(s *godog.Suite) {
...
+		fmt.Println("Running Integration test in : ", testDir)
```
Comment (praveenkumar) [inline test/integration/basic_test.go]: This will help to find where the actual test is running to findout which directory it's in.

Comment (hferentschik) [inline test/integration/basic_test.go]: +1, eventually it would be nice to have this as some configurable log output, but let's have it as message to std out for now

File: test/integration/features/basic.feature [hunk 15]
```diff
@@ -8,6 +8,17 @@ Feature: Basic
      Then Minishift should have state "Running"
       And Minishift should have a valid IP address
 
+     # Check if developer have sudo permission then oc should output as below with other
+     # cluster roles
+     # sudoer   /sudoer   developer
+     When executing "oc --as system:admin get clusterrolebindings"
```
Comment (hferentschik) [inline test/integration/features/basic.feature]: 👍 awesome

File: test/integration/util/util.go [hunk 16]
```diff
@@ -80,6 +92,13 @@ func (m *MinishiftRunner) IsRunning() bool {
 	return m.GetStatus() == "Running"
 }
 
+func (m *MinishiftRunner) GetOcRunner() *OcRunner {
+	if m.IsRunning() {
```
Comment (hferentschik) [inline test/integration/util/util.go]: :-) Right, that's what I meant

File: test/integration/util/util.go [hunk 17]
```diff
@@ -115,12 +134,10 @@ func (m *MinishiftRunner) CheckStatus(desired string) bool {
 	return m.GetStatus() == desired
 }
 
-type OcRunner struct {
-	CommandPath string
-}
-
 func NewOcRunner() *OcRunner {
-	p, _ := exec.LookPath("oc")
+	jsonDataPath := filepath.Join(os.Getenv(constants.MiniShiftHomeEnv), "machines", constants.MachineName+".json")
```
Comment (hferentschik) [inline test/integration/util/util.go]: 👍

**Comment by hferentschik** [comment]
Merged, thanks. Very nice indeed.
